### ISSUE TITLE
Questions Regarding Reproducibility and Hyperparameters

### ISSUE THREAD

**Issue by sun1187** [mala-lab/GGAD#5]
Hello, thank you for your great work!

It was fascinating to come across this kind of insight.
I’m currently working on reproducing your results, and have a few questions I’d greatly appreciate your clarification on:

1. **About hyperparameters**
    
    Are the results reported in Table 1 based on the default hyperparameter values provided in the code?
    
    If so, could you let me know the `num_epoch` and `lr` used for AEGIS’s *Ellipicit*? These are not specified in the code.
    
2. **About model selection**
    
    Could you specify whether the results in Table 1 come from the last epoch or from the one with the lowest train loss?
    
    I ran the GAD model using `run.py` and `model.py`, and performed inference at the final epoch, but the AUC result on T-Finance was around 0.5, which quite differs from the reported results(0.8). I'm not sure whether this discrepancy is due to differences in hyperparameter settings or something else.
    
3. **About negative sampling**
    
    The default value of `neg_sample` is 1, but the paper states that "the size of the generated outlier nodes S is set to 5%." Does this mean that `neg_sample` should be set to 5?
    

I’d be very grateful if you could help me better understand above questions.

Thank you very much in advance!

**Comment by fengduqianhe** [comment]
Hi sun1187， thank you so much for following our work. I would like to provide a point-to-point clarification  as the following 

1. Most of the baseline results reported in Table 1 are based on the default hyperparameters provided in their original papers. However, these default settings provided in their papers are sometimes tailored to synthetic datasets  (like Cora, ACM) and may not perform well on datasets with genuine anomalies (Amazon, T-Finance). Therefore, for such cases, we further fine-tune key hyperparameters, including the number of training epochs and learning rate, to ensure proper loss convergence.

For the AEGIS,  the lr is set to 3e-3 on elliptic, and the epoch number is set to 500.

<img width="1055" height="102" alt="Image" src="https://github.com/user-attachments/assets/024ca1ac-0b89-45f1-9f12-aaedc592427f" />

<img width="1114" height="265" alt="Image" src="https://github.com/user-attachments/assets/df4c4037-f64e-4bac-833a-5f50390abf8a" />

For the baselines, we obtain the best results by tuning the hyperparameters. In contrast, for our GGAD, we strive to use a unified of hyperparameters, which is more suitable for scenarios without labels and a validation set. 

2. For the results in Table 1,  due to the lack of a validation set and significant variation in graph data sizes, we typically categorize the datasets into small, medium, and large groups and apply a unified number of optimization steps within each category. During multiple rounds of optimization, we monitor the change in loss between consecutive rounds. When we observe that the loss has stabilized and shows little to no change, we record the corresponding result.  This stage also corresponds to the final phase of the epoch number we set. You could run the experiment and output the results every 20 epochs.

 For the GGAD performance on the T-Finance,  could you please run the run.py using the  lr=1e-3, the epoch number = 500, and the ratio of generated anomalies is 0.15, which is set in the util.py. We have run it many times, and during the optimization, the AUROC can even achieve 0.84.

<img width="1096" height="230" alt="Image" src="https://github.com/user-attachments/assets/2a70a183-8eba-41dd-95d6-1750c069f804" />

3. The neg_sample hyperparameter does not control the rate of generated anomalies; it is only related to the BCE loss function and does not require adjustment. The rate of generated anomalies is in until.py as shown in the above figure.

In summary, we sincerely appreciate your questions and attention. Please feel free to reach out if you have any further questions.

**Comment by sun1187** [comment]
Thank you very much for your prompt response. Your clarification helped me clearly understand some of the points I was previously confused about.

Regarding point 2, in the GitHub code, model.eval() is performed every 10% of the training process. After modifying it to evaluate at every epoch, I was able to verify it as expected.

If I may, I have a couple of additional questions regarding reproduction with DGraph data:

1. When I tried running python src/main.py, I encountered an error indicating that `test_pcgnn` was missing from src/utils. I was unable to locate this part in the codebase, so I would appreciate it if you could kindly confirm whether if something might be missing.

2. I'm also a bit confused about how to run the other baseline models. According to the README, it suggests executing commands like `python src/graphsage_dominant.py ...`, but the graphsage_dominant.py file appears to contain only class definitions and does not include a main function. Am I supposed to run these models through main.py by changing the handler, or is there something I might have overlooked?

I truly appreciate your kind and helpful support. Thank you once again.

**Comment by fengduqianhe** [comment]
Hi sun1187

Thank you for pointing out this issue. You are right,  I double-checked the code, and it should be 

from src.utils import **test_ggad**, test_sage, load_data, pos_neg_split, normalize

In the src/model_handle.py

Now I have updated the crosspond code, you could pull it and run it again.
<img width="935" height="341" alt="Image" src="https://github.com/user-attachments/assets/aa54ebc5-7a20-4fec-a43c-2081c1d76971" />

For the baseline,  we should change the model handle in main.py to run each baseline, since there is no main function in each sage.xx.py. I will update the readme shortly based on your suggestion. 

In summary, thank you so much for helping us refine the code, and sincerely appreciate your questions and attention. Please feel free to reach out if you have any further questions.

**Comment by sun1187** [comment]
Thank you so much for the quick response and the code fix — it really helped clear up my confusion.

After pulling the updated code and running python src/main.py, **I got an error saying that config/pcgnn_dgraph.yml is missing.**

<img width="457" height="128" alt="Image" src="https://github.com/user-attachments/assets/304907c6-c304-4cef-9638-e37c31a5e8dc" />

It seems like that particular file and parts might have been left out in the current version of the code.

Whenever you get a chance, I'd really appreciate it if you could upload that parts as well.

No rush at all — whenever it's convenient for you is totally fine.

I truly appreciate your kind and helpful responses.

**Comment by fengduqianhe** [comment]
HI sun1187
Thank you so much for pointing out ! 

This parameter should be dgraph.xml.  I may have accidentally overlooked updating main.py before pushing to GitHub.

<img width="1066" height="437" alt="Image" src="https://github.com/user-attachments/assets/e36148d5-42ac-4d5f-8457-c69933f0946a" />

Now, you may try to run the GGAD and the competing methods again  and feel free to reach out if you have any further questions.

Best
Hezhe

**Comment by sun1187** [comment]
Thank you for your prompt and thoughtful feedback — it really helped clarify my questions.

Though there were a few errors, but they have been resolved.

Great work, and I truly appreciate your support!
